### PR TITLE
Chaos Mode — Property-Based Negative Testing &amp; Unknown-Unknown Fuzzing (Closes #154)

### LINKED ISSUE #154: [INNOVATION] Chaos Mode — Property-Based Negative Testing &amp; Unknown-Unknown Fuzzing
Chaos Mode is chaos-kitten's negative testing engine. Beyond known attack signatures, it randomly generates structurally invalid inputs — wrong types, boundary extremes, null bytes, Unicode edge cases, missing required fields — to discover undocumented crashes, 500 errors, and hidden behaviors that reveal deeper vulnerabilities.

**Why this wins:** Shannon AI only tests for known OWASP categories. Chaos Mode finds the unknown unknowns — the 0-days that no payload list covers. Real-world bugs often live at the boundary between "expected input" and "unexpected but valid-looking input."

**Example findings unique to Chaos Mode:**
- `{"age": null}` → 500 Internal Server Error (unhandled null dereference)
- `{"name": "A" * 100000}` → application hangs (ReDoS)
- `{"price": 1e308}` → Float overflow causes negative balance
- Missing `Content-Type` header → XML parser fallback reveals XXE

## Key Features
- `--chaos` flag enables chaos mode alongside normal attack profiles
- Property-based test generation per field type (int: min/max/0/-1/NaN; string: empty/null/huge/Unicode/control chars; arrays: empty/nested/overflowed)
- Detect anomalies: status 500, response time > 3× baseline, response body length anomaly, server crash (connection refused)
- Configurable chaos intensity: `--chaos-level [1-5]` (1=gentle, 5=maximum carnage)
- Report chaotic findings separately: "[CHAOS] Unexpected 500 on null `userId` — potential NullReferenceException"

## Acceptance Criteria
- [ ] Create `chaos_kitten/brain/chaos_engine.py` with `ChaosEngine` class
- [ ] Implement type-aware random input generators for: int, float, string, boolean, array, object, null
- [ ] Add anomaly detectors: status 5xx, outlier response time (>3σ), connection errors
- [ ] Add `--chaos` and `--chaos-level` flags to `cli.py`
- [ ] Integrate into orchestrator as optional post-standard-scan phase
- [ ] Add `tests/test_chaos_engine.py` with 80%+ coverage
- [ ] Document chaos mode in `docs/getting_started.md`

## Apertre 3.0
**Difficulty:** Hard | **Points:** 10

To get assigned, comment below with:
```
I'd like to work on this!
Discord: [your_discord_username]
Approach: [brief description of type-aware fuzzing strategy and anomaly detection]
```

**Verification Checklist:**
- [ ] Forked the repository
- [ ] Read CONTRIBUTING.md
- [ ] Discord username matches Apertre 3.0 registration

### PR DESCRIPTION
**Pull request by vedhapprakashni**
Description:

This pull request adds Chaos Mode to chaos-kitten, a negative testing engine that goes beyond known OWASP attack signatures. Instead of only testing for documented vulnerability patterns, Chaos Mode randomly generates structurally invalid inputs to discover unknown crashes, 500 errors, and hidden behaviors that no payload list covers.

The core of this feature lives in a new file called chaos_engine.py inside the brain folder. It contains four main classes.

ChaosInput is a simple container that holds a single chaos test input along with metadata like the field name, the original expected type, the mutated value, and a human-readable description of what the input is testing.

ChaosGenerator is the type-aware random input generator. It knows how to produce chaotic inputs for each data type. For integers, it generates values like zero, negative one, Int32 minimum and maximum, very large numbers, NaN, infinity, and strings where numbers are expected. For floats, it generates negative zero, denormalized floats, overflow boundaries like 1e308, and NaN. For strings, it generates empty strings, null bytes, control characters, very long strings up to 100K characters, Unicode direction overrides, zero-width characters, emoji floods, unpaired surrogates, template injection patterns, and command injection attempts. For booleans, it sends truthy but incorrect values like the integer 2 or the string "yes". 

For arrays, it sends empty arrays, deeply nested arrays, very large arrays with 10K elements, and mixed-type arrays. For objects, it tests prototype pollution payloads, objects with 1000 keys, null bytes in key names, and wrong types entirely. It also generates missing required field payloads and chaotic HTTP headers like missing Content-Type, XML Content-Type to probe for XXE, mismatched Content-Length, and SSRF bypass headers.

The number of mutations generated per field is controlled by the chaos level, which ranges from 1 to 5. Level 1 generates only 2 mutations per field for gentle testing. Level 5 generates 12 mutations per field for maximum carnage.

AnomalyDetector analyzes the responses from chaos tests to find problems. It detects four types of anomalies. First, it catches any 5xx server errors, which indicate the server crashed or threw an unhandled exception. Second, it detects response time outliers by comparing against a baseline using the three-sigma rule, which can reveal ReDoS or algorithmic complexity attacks. Third, it scans response bodies for information leaks like stack traces, exception names, SQL errors, and file paths that should never be exposed. Fourth, it flags connection errors which could mean the server process crashed entirely.

ChaosEngine is the main orchestration class that ties everything together. It takes an endpoint definition with field names and types, generates all the chaos payloads using ChaosGenerator, sends them, and runs each response through AnomalyDetector. Findings are reported separately from standard scan results with a CHAOS prefix so they are easy to identify.

On the CLI side, two new flags have been added to the scan command. The chaos flag enables chaos mode alongside the normal attack profiles. The chaos-level flag accepts a value from 1 to 5 to control the intensity. The default level is 3.

In the orchestrator, chaos testing runs as an optional phase after the standard attack profiles complete. This means you get both the regular vulnerability scan and the chaos testing in a single run. The results dictionary now includes a chaos_findings key alongside the existing vulnerabilities key.

The test file contains 33 tests organized into five test classes covering ChaosInput, ChaosGenerator, AnomalyDetector, ChaosEngine, and AnomalyResult. All 33 tests pass. The tests verify that input generators produce the right types of mutations, that the anomaly detector correctly identifies server errors, response time outliers, and information leaks, that it does not flag normal 400 responses as anomalies, and that the chaos level properly limits the number of mutations generated.

The getting started documentation has been updated with a new section explaining how to use Chaos Mode, including example commands, the types of inputs it tests, example findings, and a table describing each chaos level.

To test this feature, run a scan with the chaos flag enabled against a target API. For example, chaos-kitten scan with the chaos flag and target set to your local API. You can increase the chaos level to 5 for more thorough testing. The output will show standard scan results first, followed by any chaos findings prefixed with CHAOS tags.

Preview

image

image

image

## Summary by CodeRabbit

* **New Features**
  * Chaos Testing mode: configurable intensity (levels 1–5) that generates invalid inputs, simulates tests, and summarizes chaos findings alongside vulnerabilities.

* **CLI**
  * scan command adds chaos and chaos_level options; simplifies run flow, reporting, and final messaging.

* **Refactor**
  * Orchestration simplified to a run-based flow that optionally includes chaos testing and returns vulnerabilities plus chaos findings.

* **Documentation**
  * "Chaos Mode (Negative Testing)" guide and chaos levels added.

* **Tests**
  * Comprehensive tests for chaos generation, detection, orchestration, and summaries.

### CODE REVIEW

File: chaos_kitten/brain/orchestrator.py [hunk 1]
```diff
@@ -190,126 +192,107 @@ class Orchestrator:
...
+        # Nodes
         workflow.add_node("recon", partial(run_recon, app_config=self.config))
         workflow.add_node("parse", parse_openapi)
         workflow.add_node("plan", plan_attacks)
-        workflow.add_node(
-            "execute_analyze", partial(execute_and_analyze, executor=executor)
-        )
+        
+        executor = Executor(self.config)
+        workflow.add_node("execute", partial(execute_and_analyze, executor=executor))
```
Comment (mdhaarishussain) [inline chaos_kitten/brain/orchestrator.py]: **🔴 Critical — `Executor` created without async context manager; `_client` will be `None` on every request.**

`Executor.__init__` does NOT create the `httpx.AsyncClient` — that only happens in `__aenter__`. Creating `Executor(self.config)` here and passing it via `partial` means `executor._client` is `None` when `execute_attack()` runs. The existing guard in `execute_attack` then returns a dummy error dict for every single attack silently.

Fix — wrap the whole graph build + `ainvoke` call in the executor's async context manager:

```python
async with Executor(self.config) as executor:
    workflow.add_node("execute", partial(execute_and_analyze, executor=executor))
    # ... rest of the graph setup ...
    app = workflow.compile()
    final_state = await app.ainvoke(initial_state)
```

File: chaos_kitten/cli.py [hunk 2]
```diff
@@ -147,93 +159,58 @@ def scan(
...
-        # Support legacy path
-        if "api" not in app_config: app_config["api"] = {}
-        app_config["api"]["spec_path"] = spec
+        # Load config or use defaults
+        try:
+            from chaos_kitten.utils.config import Config
+            config_obj = Config(config)
+            cfg = config_obj.load()
+        except Exception:
+            cfg = {"target": {}, "reporting": {}}
```
Comment (mdhaarishussain) [inline chaos_kitten/cli.py]: **🟠 Major — `except Exception` swallows the `FileNotFoundError` that guides users to run `chaos-kitten init`.**

With this guard, a user who runs `chaos-kitten scan` with no config file and no `--target` will silently scan an empty `cfg` (no base URL), get a confusing `ValueError: Target URL not configured` from the orchestrator rather than the helpful message `Config file not found: ... Run 'chaos-kitten init'`.

Catch only `FileNotFoundError` for the "no config file" case, and let everything else propagate:

```python
try:
    from chaos_kitten.utils.config import Config
    config_obj = Config(config)
    cfg = config_obj.load()
except FileNotFoundError:
    if not target and not demo and not chaos:
        console.print(f"[bold red]❌ Config file not found: {config}[/bold red]")
        console.print("Run 'chaos-kitten init' or provide --target and --spec args.")
        raise typer.Exit(code=1)
    cfg = {"target": {}, "reporting": {}}
```

**Review by mdhaarishussain — CHANGES_REQUESTED** [review]
## PR #168 — Chaos Mode (2nd Review)

**Great progress.** The 4 follow-up commits correctly fixed: duplicate `__future__` import in `openapi_parser.py`, missing `asyncio`/`json` imports in `orchestrator.py`, safe JSON serialization with `ensure_ascii=True` for the surrogate payload, and the CLI now displays scan+chaos findings. That covers all original CodeRabbit blockers.

**3 issues remain that need fixes before merge:**

### 🔴 Critical
**`Executor` not used as async context manager** (see inline comment on `orchestrator.py:244`). `httpx.AsyncClient` is initialized in `__aenter__`, not `__init__`. With the current code, `executor._client` is `None` when every attack executes — all HTTP requests silently fail with a dummy error response. Zero attacks actually reach the target. Fix: wrap the graph setup and `ainvoke` call inside `async with Executor(self.config) as executor:`.

### 🟠 Major  
**Blind `except Exception` in `cli.py` swallows helpful `FileNotFoundError`** (see inline comment on `cli.py:177`). Change to `except FileNotFoundError` and show the original guidance message.

### 🟡 Minor
**Chaos mode silently returns `[]` when LangGraph is unavailable** (see inline comment on `orchestrator.py:233`), even though `ChaosEngine` has no LangGraph dependency. Split the LangGraph guard so only the graph-based standard scan bails out, letting `--chaos` still run.

Once these 3 are fixed this PR can be approved.

**Review by mdhaarishussain — CHANGES_REQUESTED** [review]
## Review — Round 2 (9 commits)

The `from __future__ import annotations` placement fix is confirmed. Three issues remain.

---

### 🔴 Critical — `Executor(self.config)` passes config dict instead of `base_url` string

**`orchestrator.py` – `run()`, ~line 354:**

```python
async with Executor(self.config) as executor:
```

`Executor.__init__` expects `base_url: str` as its first argument. Passing the entire config dict causes a `TypeError` at runtime — every scan fails. Fix:

```python
async with Executor(self.config.get("target", {}).get("base_url")) as executor:
```

---

### 🟠 Major — `--provider` flag is silently ignored

**`cli.py` – config override block:**

```python
cfg = {"target": {}, "reporting": {}}
if target:
    cfg["target"]["base_url"] = target
if spec:
    cfg["target"]["openapi_spec"] = spec
if output:
    cfg["reporting"]["output_path"] = output
# provider override is missing!
```

`execute_and_analyze` reads `cfg["agent"]["llm_provider"]` to pick the LLM. If a user passes `--provider openai`, the CLI silently ignores it and falls back to whatever default is in the config. Add:

```python
if provider:
    cfg.setdefault("agent", {})["llm_provider"] = provider
```

---

### 🟠 Major — `except Exception` swallows `FileNotFoundError` — unfriendly error for missing config

**`cli.py` – config loading block:**

```python
try:
    from chaos_kitten.brain.orchestrator import Orchestrator
    import asyncio
    cfg = {...}
    ...
except Exception as e:
    ...
```

The owner already flagged this: catching `Exception` means `FileNotFoundError` (config file missing) is silently swallowed instead of showing the helpful "Run `chaos-kitten init`" message. Narrow the exception to only catch `FileNotFoundError` for the config-missing case and let everything else propagate.

---

### Summary

| # | Severity | Issue |
|---|----------|-------|
| 1 | 🔴 Critical | `Executor(self.config)` — must pass `base_url: str`, not the config dict |
| 2 | 🟠 Major | `--provider` CLI flag silently ignored in config override block |
| 3 | 🟠 Major | `except Exception` swallows `FileNotFoundError` from config load |

Please fix all three before merging.

**Review by mdhaarishussain — CHANGES_REQUESTED** [review]
## ❌ Request Changes — Round 4

Excellent progress — all four issues from Round 3 are now fixed:

- ✅ `payload_obj` → `payload_val` (line 176)
- ✅ `partial(execute_and_analyze, executor=executor, app_config=self.config)` (line 356)
- ✅ `except FileNotFoundError` guard properly isolated
- ✅ `--format` wired into `cfg` via `cfg.setdefault("reporting", {})["format"] = format`

**Two new regressions introduced in this round:**

---

### 🔴 Critical — Report generation removed, replaced with a stub print

`cli.py` lines 215–216:
```python
# Generate report (mock for now)
console.print(f"\n📝 Generating {format} report in {output}...")
```
The actual `Reporter.generate()` call that existed in the previous version has been deleted and replaced with a print statement that does nothing. Scans complete but **no report file is ever written**. If `format` or `output` are `None` (their CLI defaults), this also prints `"Generating None report in None..."` which is confusing.

**Fix** — bring back report generation:
```python
reporter_cfg = cfg.get("reporting", {})
reporter = Reporter(
    output_path=reporter_cfg.get("output_path", "./reports"),
    output_format=reporter_cfg.get("format", "html"),
)
report_file = reporter.generate({"vulnerabilities": results.get("vulnerabilities", [])}, target_url)
console.print(f"\n📄 Report: [underline]{report_file}[/underline]")
```

---

### 🟠 Major — `--fail-on-critical` logic removed

The `fail_on_critical` CLI flag and its enforcement block were deleted from `cli.py`. This flag is documented in the README and relied upon in CI pipelines to fail the build when critical vulnerabilities are found. Silently removing it is a breaking change.

**Fix** — restore the check after scanning:
```python
if fail_on_critical:
    critical_vulns = [v for v in vulnerabilities if str(v.get("severity", "")).lower() == "critical"]
    if critical_vulns:
        console.print(f"[bold red]❌ {len(critical_vulns)} critical vulnerabilities. Failing pipeline.[/bold red]")
        raise typer.Exit(code=1)
```

---

Please restore these two behaviours before this can merge.

**Review by mdhaarishussain — CHANGES_REQUESTED** [review]
## Chaos Mode — Negative Testing & Fuzzing — Round 6 Review

### ✅ Confirmed Fixed From the Diff

- **`execute_and_analyze` + `app_config`** → `partial(execute_and_analyze, executor=executor, app_config=self.config)` ✅ (line 434)
- **`target_url` defined in `run()`** → `self.config.get("target", {}).get("base_url")` ✅ (no more NameError)
- **`nl_plan` cleanup** → removed from both graph and `initial_state` correctly ✅ (matches de-scoping of the `--goal` feature)
- **Executor as async context manager** → `async with Executor(...) as executor:` ✅
- **reporter.generate() and --fail-on enforcement** → Credited as fixed per commits `791465e` and `9212266`; coderabbitai found no new issues on those paths

---

### 🟠 Major — `parse_openapi` and `plan_attacks` nodes not bound to `app_config` (orchestrator.py L424–426)

```python
workflow.add_node("recon",  partial(run_recon, app_config=self.config))  # ✅ correct
workflow.add_node("parse",  parse_openapi)                               # ← no partial
workflow.add_node("plan",   plan_attacks)                                # ← no partial
workflow.add_node("execute", partial(execute_and_analyze, executor=executor, app_config=self.config))  # ✅ correct
```

`run_recon` and `execute_and_analyze` both receive `app_config` via `partial`. LangGraph will call `parse_openapi` and `plan_attacks` with only the state dict — if those functions declare an `app_config` parameter (as the node pattern across the codebase implies), they'll receive `None` and silently fall back to defaults. Wrap them consistently:

```python
workflow.add_node("parse", partial(parse_openapi, app_config=self.config))
workflow.add_node("plan",  partial(plan_attacks,  app_config=self.config))
```

If those functions genuinely don't accept `app_config`, confirm that with an inline comment explaining why they don't need it.

---

### ✅ Good

- Chaos engine architecture is clean and well-structured
- Tests for `ChaosGenerator`, `AnomalyDetector`, and `ChaosEngine` are comprehensive
- Level-clamping logic and graceful fallback when LangGraph is absent are both solid
